fairlearn/_input_validation.py

```diff
@@ -3,91 +3,88 @@
         raise ValueError(_MESSAGE_Y_NONE)
-
-    if _KW_SENSITIVE_FEATURES not in kwargs:
-        msg = "Must specify {0} (for now)".format(_KW_SENSITIVE_FEATURES)
-        raise RuntimeError(msg)
-
-    # Extract the target attribute
-    sensitive_features_vector = _make_vector(kwargs[_KW_SENSITIVE_FEATURES],
-                                             _KW_SENSITIVE_FEATURES)
-
-    if enforce_binary_sensitive_feature:
-        unique_labels = np.unique(sensitive_features_vector)
-        if len(unique_labels) > 2:
-            raise RuntimeError("Sensitive features contain more than two unique values")
-
-    # Extract the Y values
-    y_vector = _make_vector(y, "y")
-
-    X_rows, _ = _get_matrix_shape(X, "X")
-    if X_rows != y_vector.shape[0]:
-        raise RuntimeError(_MESSAGE_X_Y_ROWS)
-    if X_rows != sensitive_features_vector.shape[0]:
-        raise RuntimeError(_MESSAGE_X_SENSITIVE_ROWS)
-
-    return pd.DataFrame(X), y_vector, sensitive_features_vector
-
-
-def _make_vector(formless, formless_name):
-    formed_vector = None
-    if isinstance(formless, list):
-        formed_vector = pd.Series(formless)
-    elif isinstance(formless, pd.DataFrame):
-        if len(formless.columns) == 1:
-            formed_vector = formless.iloc[:, 0]
-        else:
-            msgfmt = "{0} is a DataFrame with more than one column"
-            raise RuntimeError(msgfmt.format(formless_name))
-    elif isinstance(formless, pd.Series):
-        formed_vector = formless
-    elif isinstance(formless, np.ndarray):
-        if len(formless.shape) == 1:
-            formed_vector = pd.Series(formless)
-        elif len(formless.shape) == 2 and formless.shape[1] == 1:
-            formed_vector = pd.Series(formless[:, 0])
-        else:
-            msgfmt = "{0} is an ndarray with more than one column"
-            raise RuntimeError(msgfmt.format(formless_name))
     else:
-        msgfmt = "{0} not an ndarray, Series or DataFrame"
-        raise RuntimeError(msgfmt.format(formless_name))
-
-    return formed_vector
-
+        X = check_array(X)
+
+    sensitive_features = kwargs.get(_KW_SENSITIVE_FEATURES)
+    if sensitive_features is None:
+        raise ValueError(_MESSAGE_SENSITIVE_FEATURES_NONE)
+
+    check_consistent_length(X, sensitive_features)
+    sensitive_features = check_array(sensitive_features, ensure_2d=False, dtype=None)
+
+    # compress multiple sensitive features into a single column
+    if len(sensitive_features.shape) > 1 and sensitive_features.shape[1] > 1:
+        sensitive_features = np.apply_along_axis(
+                lambda row: _SENSITIVE_FEATURE_COMPRESSION_SEPARATOR.join(
+                    [str(row[i])
+                     .replace("\\", "\\\\")  # escape backslash and separator
+                     .replace(_SENSITIVE_FEATURE_COMPRESSION_SEPARATOR,
+                              "\\" + _SENSITIVE_FEATURE_COMPRESSION_SEPARATOR)
+                     for i in range(len(row))]),
+                axis=1,
```

Comment:
I know it's a small piece of code, but I wonder if we can pull out the "string joining" functionality into a little utility function, so you could directly pass it to the apply function. This will make it easier if in future we want to apply the same strategy to the column names.

Reply:
we could also make it a function in the future if it's needed again ;)

Reply:
It may also be needed in widget code, although I don't know details there. I'll make it its own function

Reply:
done in the other PR!